### PR TITLE
SCEV: return std::nullopt for invalid TC (NFC)

### PR DESCRIPTION
**Pull request by artagnon**
ScalarEvolution::getSmallConstantTripCount and getSmallConstantMaxTripCount have a special zero return value to indicate that the trip count is invalid (unknown, too large, or unsigned-wrap). This can cause confusion in callers. Change it to never wrap, and return an std::optional that has a value on valid trip counts. This purpose of this patch is to improve the API, and it is non-functional.

### CODE REVIEW

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 1]
```diff
@@ -6463,7 +6463,7 @@ getRangeForUnknownRecurrence(const SCEVUnknown *U) {
     // TODO: Handle the power function forms some day.
     return FullSet;
 
-  unsigned TC = getSmallConstantMaxTripCount(L);
+  std::optional<unsigned> TC = getSmallConstantMaxTripCount(L);
   if (!TC || TC >= BitWidth)
```
Comment (tschuett) [inline llvm/lib/Analysis/ScalarEvolution.cpp:6467]:
```
TC >= BitWidth
```
It compiles? We compare an optional to a BitWidth?

**Comment by nikic** [comment]
The PR description (but especially the title) are confused. On wrap, the TC is treated as unknown, which is represented as a 0 value. There was no "conflation" here.

> Improvements are planned for future patches.

What improvements are you planning?

File: llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp [hunk 2]
```diff
@@ -6822,8 +6822,8 @@ canFoldTermCondOfLoop(Loop *L, ScalarEvolution &SE, DominatorTree &DT,
   // the allowed cost with the loops trip count as best we can.
   const unsigned ExpansionBudget = [&]() {
     unsigned Budget = 2 * SCEVCheapExpansionBudget;
-    if (unsigned SmallTC = SE.getSmallConstantMaxTripCount(L))
-      return std::min(Budget, SmallTC);
+    if (std::optional<unsigned> SmallTC = SE.getSmallConstantMaxTripCount(L))
+      return std::min(Budget, *SmallTC);
```
Comment (tschuett) [inline llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:6826]: Is the `*` really needed below the`if` ?

**Comment by artagnon** [comment]
> The PR description (but especially the title) are confused. On wrap, the TC is treated as unknown, which is represented as a 0 value. There was no "conflation" here.

Yes, sorry. Title fixed.

> > Improvements are planned for future patches.
> 
> What improvements are you planning?

The patch was motivated by LV printing a "Found TC: 0".

File: llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp (hunk 2 above)
Comment (artagnon) [reply llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:6826]: Yes.

File: llvm/lib/Analysis/ScalarEvolution.cpp (hunk 1 above)
Comment (artagnon) [reply llvm/lib/Analysis/ScalarEvolution.cpp:6467]: See std::optional<T>::operator>=.

File: llvm/lib/Analysis/LoopCacheAnalysis.cpp [hunk 3]
```diff
@@ -568,9 +568,10 @@ CacheCost::CacheCost(const LoopVectorTy &Loops, const LoopInfo &LI,
   assert(!Loops.empty() && "Expecting a non-empty loop vector.");
 
   for (const Loop *L : Loops) {
-    unsigned TripCount = SE.getSmallConstantTripCount(L);
-    TripCount = (TripCount == 0) ? DefaultTripCount : TripCount;
-    TripCounts.push_back({L, TripCount});
+    std::optional<unsigned> TripCount = SE.getSmallConstantTripCount(L);
+    if (!TripCount)
+      TripCount = DefaultTripCount;
```
Comment (nikic) [inline llvm/lib/Analysis/LoopCacheAnalysis.cpp:573]:
```suggestion
    unsigned TripCount = SE.getSmallConstantTripCount(L).value_or(DefaultTripCount);
```

File: llvm/lib/Target/Hexagon/HexagonTargetTransformInfo.cpp [hunk 4]
```diff
@@ -88,9 +88,8 @@ void HexagonTTIImpl::getPeelingPreferences(Loop *L, ScalarEvolution &SE,
                                            TTI::PeelingPreferences &PP) {
   BaseT::getPeelingPreferences(L, SE, PP);
   // Only try to peel innermost loops with small runtime trip counts.
-  if (L && L->isInnermost() && canPeel(L) &&
-      SE.getSmallConstantTripCount(L) == 0 &&
-      SE.getSmallConstantMaxTripCount(L) > 0 &&
+  if (L && L->isInnermost() && canPeel(L) && !SE.getSmallConstantTripCount(L) &&
+      SE.getSmallConstantMaxTripCount(L) &&
       SE.getSmallConstantMaxTripCount(L) <= 5) {
```
Comment (nikic) [inline llvm/lib/Target/Hexagon/HexagonTargetTransformInfo.cpp:93]:
```suggestion
      SE.getSmallConstantMaxTripCount(L) <= 5) {
```
I think that should work now? Calling the method twice is not great...

File: llvm/lib/Target/PowerPC/PPCTargetTransformInfo.cpp [hunk 5]
```diff
@@ -346,7 +346,7 @@ bool PPCTTIImpl::isHardwareLoopProfitable(Loop *L, ScalarEvolution &SE,
   SchedModel.init(ST);
 
   // Do not convert small short loops to CTR loop.
-  unsigned ConstTripCount = SE.getSmallConstantTripCount(L);
+  unsigned ConstTripCount = SE.getSmallConstantTripCount(L).value_or(0);
```
Comment (nikic) [inline llvm/lib/Target/PowerPC/PPCTargetTransformInfo.cpp:349]: value_or doesn't seem useful here, given the only uses on the line below.

**Comment by nikic** [comment]
> If the only planned use of this is to avoid printing `TC: 0` in LV, would it be possible to simply update the code in LV to check that the return value is != 0 before printing? IIUC 0 is never a valid return value for `getSmallConstantTripCount` and it should be fairly clear from the documentation.

Also, just to double check, is that `TC: 0` print really a bug and not intentional? I would have thought it is relevant debug information that the trip count could not be computed.

**Comment by artagnon** [comment]
> Also, just to double check, is that `TC: 0` print really a bug and not intentional? I would have thought it is relevant debug information that the trip count could not be computed.

If it's desired, we should print "Could not compute trip count", as opposed to a cryptic "Found TC: 0".

**Comment by artagnon** [comment]
> If the only planned use of this is to avoid printing `TC: 0` in LV, would it be possible to simply update the code in LV to check that the return value is != 0 before printing? IIUC 0 is never a valid return value for `getSmallConstantTripCount` and it should be fairly clear from the documentation.

Wouldn't it be preferable if this was clear from the API? I'm also planning to change the API of related functions like `getTripCountFromExitCount` so that there isn't an ugly and cryptic `TC->isZero()` check.

File: llvm/lib/Target/Hexagon/HexagonTargetTransformInfo.cpp (hunk 4 above)
Comment (artagnon) [reply llvm/lib/Target/Hexagon/HexagonTargetTransformInfo.cpp:93]: `std::optional<T>::operator<=(const optional<T>& opt, const T& value)` returns true when `opt` is `std::nullopt`, so this won't work.

**Comment by artagnon** [comment]
Gentle ping. This patch is just a nice-to-have for clarifying the API.

**Comment by artagnon** [comment]
Gentle ping.

**Comment by artagnon** [comment]
Not pursuing this.

**Pull request closed without merging**